### PR TITLE
Add apartment#index & property#show

### PR DESCRIPTION
**Pull request by ricardoruwer**
* Change icons to SVG
* Rearrange some template files

---

A lot of changes 😆

### CODE REVIEW

File: lib/construtora_lc_hiert/properties.ex [hunk 1]
```diff
@@ -51,9 +68,21 @@ defmodule ConstrutoraLcHiert.Properties do
...
   """
-  def list_properties do
-    Repo.all(Property)
+  def list_properties(), do: Repo.all(Property)
+
+  def list_properties(type: type) do
+    Repo.all(from p in Property, where: p.type == ^type)
+  end
+
+  def list_featured_properties(limit: limit) do
```
Comment (ricardoruwer) [inline lib/construtora_lc_hiert/properties.ex]: So, I have 3 functions here:
- `list_properties()`
It lists all properties (houses, apartments and lots) and I'm currently using it only in the admin page to display all the records.

- `list_properties(type: type)`
It lists all the properties given a type, for example `:apartment`, and I'm using it to list the properties on the 3 pages I have on the website.
image

- `list_featured_properties(limit: limit)`
It lists the featured properties, and it has a `limit`, for example, 6 in the `home#index` page and 2 in the `property#show` page. And it also has the `order_by: [desc: p.inserted_at]`.

Do you think in a better way to do it all?

File: lib/construtora_lc_hiert_web/helpers/icon_helper.ex [hunk 2]
```diff
@@ -3,9 +3,14 @@ defmodule ConstrutoraLcHiertWeb.Helpers.IconHelper do
   Give some icons to be used on templates.
   """
 
-  use Phoenix.HTML, only: [img_tag: 1]
+  def icon(name, opts \\ %{}) do
```
Comment (ricardoruwer) [inline lib/construtora_lc_hiert_web/helpers/icon_helper.ex]: It looks good? 😆

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,26 +1,22 @@
...
-      # specify any bash command here prefixed with `run: `
       - run: mix local.hex --force
       - run: mix local.rebar --force
       - run: mix deps.get
       - run: mix ecto.create
+
+      - run: curl -sSL "https://nodejs.org/dist/v6.17.1/node-v6.17.1-linux-x64.tar.xz" | sudo tar --strip-components=2 -xJ -C /usr/local/bin/ node-v6.17.1-linux-x64/bin/node
+      - run: curl https://www.npmjs.com/install.sh | sudo bash
+      - run: cd assets && npm install && npm run deploy && cd -
+      - run: mix phx.digest
```
Comment (ricardoruwer) [inline .circleci/config.yml]: That's the only way I could make it work. 
I need it because of the `icon_helper.ex` at 
```ex
File.read!("priv/static/images/icons/#{name}.svg")
```
So, I need to run `mix phx.digest` to have these files at `priv/static`.

File: lib/construtora_lc_hiert_web/views/property_view.ex [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  defp address(apartment) do
+    [apartment.address, apartment.address_number, apartment.complement]
+    |> Enum.reject(&is_nil/1)
+    |> Enum.join(", ")
+  end
+
+  defp city(apartment) do
+    Enum.join([apartment.city, apartment.state], "-")
+  end
+end
```
Comment (ricardoruwer) [inline lib/construtora_lc_hiert_web/views/property_view.ex]: 👀

File: lib/construtora_lc_hiert_web/router.ex [hunk 5]
```diff
@@ -29,6 +29,12 @@ defmodule ConstrutoraLcHiertWeb.Router do
     get "/quem-somos", AboutController, :index
     resources "/contato", ContactController, only: [:index, :create]
 
+    scope "/imoveis" do
+      get "/apartamentos", ApartmentController, :index
+
+      get "/:slug", PropertyController, :show
+    end
```
Comment (ricardoruwer) [inline lib/construtora_lc_hiert_web/router.ex]: 👀

File: lib/construtora_lc_hiert_web/helpers/icon_helper.ex (hunk 2 above)
Comment (andrielfn) [reply lib/construtora_lc_hiert_web/helpers/icon_helper.ex]: I'd say... no. 😺 

The IO from the `File.read!` is probably something bad here. One approach that I've been using in the last years is to use a SVG sprite instead. Basically, I have a single `icons.svg` file where I place all the SVGs and then I link them with something like the following in the HTML:

```svg
<svg class="mr-4 text-5xl text-gray-500 icon">
   <use xlink:href="/images/icons.svg#beluga"></use>
</svg>
```

As you can see, the `beluga` is the icon that I'm looking for in the sprite. And, how does it works?

```svg
<?xml version="1.0" encoding="utf-8"?>
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
  <svg id="search" viewBox="0 0 48 48">
    <path d="M31 28h-1.59l-.55-.55C30.82 25.18 32 22.23 32 19c0-7.18-5.82-13-13-13S6 11.82"/>
    <path d="M0 0h48v48H0z" fill="none"/>
  </svg>

  <svg id="chevron-right" viewBox="0 0 36 36">
    <path d="M15 9l-2.12 2.12L19.76 18l-6.88 6.88L15 27l9-9z"/>
    <path d="M0 0h36v36H0z" fill="none"/>
  </svg>

  # ...
</svg>
```

In the example above, I've placed two icons (`search` and `chevron-right`). The ID that you set in the inner SVGs are the value that you will use in the HTML, for example `<use xlink:href="/images/icons.svg#search"></use>`. However, there are some caveats that you might be aware: you must have the `viewBox` attribute of the SVGs, and if I properly remember, you need to remove the `width` and `height` that usually comes together. What I do every time is to compress the SVGs using the [SVGOMG](https://jakearchibald.github.io/svgomg/). So, compress the SVG, copy the result, remove the `width` and `height`, keep the `viewBox`, set and ID with the icon name, and you might be good.

In the side of the application, you are free to implement what you think that suits better to you. In the Elixir world, I ended up with the following:

```elixir
defmodule GoBikesWeb.IconHelpers do
  use Phoenix.HTML
  alias GoBikesWeb.Router.Helpers, as: Routes

  def icon_tag(conn, name, opts \\ []) do
    classes = Keyword.get(opts, :class, "") <> " icon"

    content_tag(:svg, class: classes) do
      tag(:use, "xlink:href": Routes.static_path(conn, "/images/icons.svg#" <> name))
    end
  end
end
```

And use it with:

```
<%= icon_tag(@conn, "search", class: "") %>
```

And there is another trick. As you can see, for every icon I append a `icon` class, where the definition is:

```css
.icon {
  /* Allows sizing by changing the icon’s font-size. */
  width: 1em;
  height: 1em;

  /* Nice visual alignment for icons alongside text. */
  vertical-align: -.15em;

   /* Default path fill = value of the color property. */
  fill: currentColor;

  /* Overwrite TailwindCSS. */
  display: inline-block;
}
```

This will make you life much easier. Basically, if your design system is properly set (i.e. 1em = base font-size), you can synchronize the icon and text size when you set the `font-size` in a parent element, for example. The `fill: currentColor` is another miracle in the technology. This will allows you to change the icon color with CSS the same way you do with font size.

In practice, you will have something like:

```
<%= link to: Routes.admin_bike_path(@conn, :new), class: "btn btn-green flex items-center" do %>
  <%= icon_tag(@conn, "add", class: "text-xl mr-1") %> Cadastrar bike
<% end %>
```

Basically, in this case, the `.btn` sets the `text-color` to white, which will ALSO make the icon color white. By default, the icon inherits the font size set in the `.btn` as well, however, as I want to make the icon a little bit bigger than the text, I've added a `text-xl` which is basically a font size definition in the CSS.

This specific scenario gave me the following:

image

I might be missing some details, but you will be fine! 😃

Comment (andrielfn) [reply lib/construtora_lc_hiert_web/helpers/icon_helper.ex]: Congratulations! You've just won a Rails versions. Get it for free:

```ruby
module ApplicationHelper
  def icon_tag(icon, options = {})
    options[:class] = Array(options[:class]) + ['icon']

    options['aria-hidden'] = true

    tag.svg options do
      tag.use 'xlink:href': "#{asset_url('icons.svg')}##{icon}"
    end
  end
end
```

File: lib/construtora_lc_hiert/properties.ex (hunk 1 above)
Comment (w-sanches) [reply lib/construtora_lc_hiert/properties.ex]: Those "named parameters" (is this what they are called? IDK) seem unnecessary. I would rather have then as simple parameters, but it is just a bit of a personal opinion here.

Also, what happens if I don't pass a limit to `list_featured_properties`? Do you think a default limit is beneficial?

Any reason the featured are ordered by `inserted_at` instead of `updated_at`? This is a business logic, obviously, so I think there is no right or wrong here. I am curious about it :p

File: lib/construtora_lc_hiert_web/views/property_view.ex (hunk 4 above)
Comment (w-sanches) [reply lib/construtora_lc_hiert_web/views/property_view.ex]: Since the `city/1` doesn't check for `nil`s I think we could drop the join and use a simple string interpolation here. WDYT?

File: lib/construtora_lc_hiert_web/router.ex (hunk 5 above)
Comment (w-sanches) [reply lib/construtora_lc_hiert_web/router.ex]: What are we 👀ing here exactly? 😆

File: lib/construtora_lc_hiert/properties.ex (hunk 1 above)
Comment (ricardoruwer) [reply lib/construtora_lc_hiert/properties.ex]: I agree with the "keyword arguments"... 😆 

I don't have any reason yet define the order of featured properties. Probably I'll change something here later...
